### PR TITLE
feat(constructor): use a strategy to have consistent behavior for constructor

### PR DESCRIPTION
**Pull request by joelwurtz**
This fix https://github.com/jolicode/automapper/issues/8

It introduces a new "way" to handle constructor : 

 * auto : will try to use constructor when possible
 * never : same as "allowConstructor: false" as before
 * always : always use the target constructor, this force user to add default value if needed in the context (otherwise il will throw an exception)

### CODE REVIEW

File: src/Configuration.php [hunk 1]
```diff
@@ -6,6 +6,10 @@
 
 final readonly class Configuration
 {
+    public const CONSTRUCTOR_STRATEGY_AUTO = 'auto';
+    public const CONSTRUCTOR_STRATEGY_ALWAYS = 'always';
+    public const CONSTRUCTOR_STRATEGY_NEVER = 'never';
```
Comment (Korbeil) [inline src/Configuration.php:11]: I would prefer us using an Enum, what do you think about it ?

File: src/Generator/CreateTargetStatementsGenerator.php [hunk 2]
```diff
@@ -195,21 +170,17 @@ private function constructorArguments(GeneratorMetadata $metadata): array
...
-        }
-
         $variableRegistry = $metadata->variableRegistry;
         $constructVar = $variableRegistry->getVariableWithUniqueName('constructArg');
         $fieldValueExpr = $propertyMetadata->source->accessor?->getExpression($variableRegistry->getSourceInput());
 
         if (null === $fieldValueExpr) {
             if (!($propertyMetadata->transformer instanceof AllowNullValueTransformerInterface)) {
-                return null;
+                return [null, null, null];
```
Comment (Korbeil) [inline src/Generator/CreateTargetStatementsGenerator.php:183]: A bit cursed, why this change ? `null` would be still fine there no ?

Comment (joelwurtz) [reply src/Generator/CreateTargetStatementsGenerator.php:183]: Don't change really anything, and phpstan works better that way, so i made it like this

File: src/Generator/CreateTargetStatementsGenerator.php [hunk 3]
```diff
@@ -218,6 +189,24 @@ private function constructorArgument(GeneratorMetadata $metadata, PropertyMetada
         /* Get extract and transform statements for this property */
         [$output, $propStatements] = $propertyMetadata->transformer->transform($fieldValueExpr, $constructVar, $propertyMetadata, $variableRegistry->getUniqueVariableScope(), $variableRegistry->getSourceInput());
 
+        $defaultValueExpr = new Expr\Throw_(
```
Comment (nikophil) [inline src/Generator/CreateTargetStatementsGenerator.php:192]: question: why don't we throw like... now? we know the mapping process will create an error. And it has the advantage to occur sooner, moreover if the mappers generation occur at warmup time

File: src/Metadata/GeneratorMetadata.php [hunk 4]
```diff
@@ -68,10 +69,14 @@ public function isSourceUserDefined(): bool
 
     public function hasConstructor(): bool
     {
-        if (!$this->allowConstructor) {
+        if ($this->constructorStrategy === ConstructorStrategy::NEVER) {
```
Comment (nikophil) [inline src/Metadata/GeneratorMetadata.php:72]: in which case would we want to never use the constructor? if the target class has a constructor, wouldn't it break?

File: docs/bundle/configuration.md [hunk 5]
```diff
@@ -9,7 +9,7 @@ Then configure the bundle to your needs, for example:
 ```yaml
 automapper:
   class_prefix: "Symfony_Mapper_"
-  allow_constructor: true
+  constructor_strategy: 'auto'
```
Comment (nikophil) [inline docs/bundle/configuration.md:12]: shouldn't we allow to configure this at runtime, thanks to the context?

File: src/Generator/CreateTargetStatementsGenerator.php (hunk 3 above)
Comment (joelwurtz) [reply src/Generator/CreateTargetStatementsGenerator.php:192]: The mapping will not throw if you provide default constructor argument in the context, there may be case where this is needed

File: src/Metadata/GeneratorMetadata.php (hunk 4 above)
Comment (joelwurtz) [reply src/Metadata/GeneratorMetadata.php:72]: Depends on what you use your mapping for, by example doctrine never use the constructor when instantating an entity from the database. At least having this, ensure a consistent behavior there is no "magic"

File: src/Generator/CreateTargetStatementsGenerator.php (hunk 3 above)
Comment (nikophil) [reply src/Generator/CreateTargetStatementsGenerator.php:192]: ok ineed! I knew I was missing something :sweat_smile:

File: docs/bundle/configuration.md (hunk 5 above)
Comment (joelwurtz) [reply docs/bundle/configuration.md:12]: I think having this at runtime would loose performance over a "niche" feature, my goal with that PR is more : I want to generate mapping with the same consistent behavior (never use constructor, always use it, try to use it if possible)

Having this at runtime will complicate thing IMO, let's wait for someone who need that to make this change if needed ?

**Review by nikophil — APPROVED** [review]
